README.md
# dSFMTSearchParam
parameter search of dSFMT


ChangeLog
2017-02-28  Mutsuo Saito <saito@manieth.com>

	* tinymtX64search.hpp for MTToolBox ver 1.0
	* AlgorithmRange.hpp for MTToolBox ver 1.0
	* tinymtX64dc.cpp for MTToolBox ver 1.0
	* README.md add requirement

2016-10-04  Mutsuo Saito <saito@manieth.com>

	* README.md

2017-01-12  Mutsuo Saito <saito@manieth.com>

	* Makefile.am remove src
	* develop/Makefile.am add sample
	* develop/sample.cpp add


src/search-test.sh
#!/bin/bash
#PBS -q q-test
###PBS -l select=1:ncpus=36::mpiprocs=36:mem=360gb
#PBS -l select=1:ncpus=36:mpiprocs=18:mem=360gb
#PBS -l walltime=01:00:00
#PBS -N dSFMTSearch
###PBS -o search_test.out.txt 不要か
#PBS -j oe
#PBS -m abe
#PBS -M sai10@hiroshima-u.ac.jp

## 自動並列化でこれを指定せよとの記述あり gcc ならなし
## export OMP_NUM_THREADS=6

### mpiexec ではいらない　dplace では必須
### export KMP_AFFINITY=disabled

cd $PBS_O_WORKDIR # 現ディレクトリに変更 ほぼ必須

### openMP/MPI ハイブリッドでは omplace を指定せよとのこと。（たぶん自動並列化も）
### gcc で openMP を使わないのでなし
mpiexec_mpt ./dSFMTdc --file dSFMTdc.521 --fixed-sl1 25 --fixed-pos1 3 --seed 100 --count 100 521


src/dSFMTsearch.hpp
#ifndef DSFMTSEARCH_HPP
#define DSFMTSEARCH_HPP
/**
 * @file dSFMTsearch.hpp
 *
 * @brief SIMD oriented Fast Mersenne Twister
 * this class is used by dSFMTdc.
 *
 * This file is important. Users should not change this file,
 * except they are experts in random number generation.
 * This file is used for parameter searching.
 *
 * @author Mutsuo Saito (Manieth Corp.)
 * @author Makoto Matsumoto (The University of Tokyo)
 *
 * Copyright (C) 2015 Mutsuo Saito, Makoto Matsumoto,
 * Manieth Corp. and Hiroshima University.
 * All rights reserved.
 *
 * The 3-clause BSD License is applied to this software, see
 * LICENSE.txt
 */

#include <iostream>
#include <iomanip>
#include <cstdlib>
#include <sstream>
#include <MTToolBox/ReducibleGenerator.hpp>
#include <MTToolBox/MersenneTwister64.hpp>
#include <MTToolBox/util.hpp>
#include "w128.hpp"

/**
 * @namespace dSFMT
 * name space for dSFMT
 */
namespace MTToolBox {
    using namespace NTL;
    using namespace std;

    /**
     * @class dSFMT_param
     * @brief a class keeping parameters of dSFMT
     *
     * This class keeps parameters of dSFMT, and has some
     * method for outputting parameters.
     */
    class dSFMT_param {
    public:
        int mexp;
        int pos1;
        int sl1;
        uint64_t msk1;
        uint64_t msk2;
        uint64_t fix1;
        uint64_t fix2;
        uint64_t parity1;
        uint64_t parity2;

        dSFMT_param() {
            mexp = 0;
            pos1 = 0;
            sl1 = 0;
            msk1 = 0;
            msk2 = 0;
            fix1 = 0;
            fix2 = 0;
            parity1 = 0;
            parity2 = 0;
        }

        dSFMT_param(const dSFMT_param& src) {
            mexp = src.mexp;
            pos1 = src.pos1;
            sl1 = src.sl1;
            msk1 = src.msk1;
            msk2 = src.msk2;
            fix1 = src.fix1;
            fix2 = src.fix2;
            parity1 = src.parity1;
            parity2 = src.parity2;
        }
        /**
         * This method is used in output.hpp.
         * @return header line of output.
         */
        const string get_header() const {
            return "mexp, pos1, sl1, msk1, msk2, fix1, fix2, parity1, parity2";
        }

        /**
         * This method is used in output.hpp.
         * @return string of parameters
         */
        const string get_string() const {
            stringstream ss;
            ss << dec << mexp << ",";
            ss << dec << pos1 << ",";
            ss << dec << sl1 << ",";
            ss << hex << setw(16) << setfill('0') << msk1 << ",";
            ss << hex << setw(16) << setfill('0') << msk2 << ",";
            ss << hex << setw(16) << setfill('0') << fix1 << ",";
            ss << hex << setw(16) << setfill('0') << fix2 << ",";
            ss << hex << setw(16) << setfill('0') << parity1 << ",";
            ss << hex << setw(16) << setfill('0') << parity2 << ",";
            string s;
            ss >> s;
            return s;
        }

        /**
         * This method is used for DEBUG.
         * @return string of parameters.
         */
        const string get_debug_string() const {
            stringstream ss;
            ss << "mexp:" << dec << mexp << endl;
            ss << "pos1:" << dec << pos1 << endl;
            ss << "sl1:" << dec << sl1 << endl;
            ss << "msk1:" << hex << setw(16) << setfill('0') << msk1 << endl;
            ss << "msk2:" << hex << setw(16) << setfill('0') << msk2 << endl;
            ss << "fix1:" << hex << setw(16) << setfill('0') << fix1 << endl;
            ss << "fix2:" << hex << setw(16) << setfill('0') << fix2 << endl;
            ss << "parity1:" << hex << setw(16) << setfill('0') << parity1
               << endl;
            ss << "parity2:" << hex << setw(16) << setfill('0') << parity2
               << endl;
            string s;
            ss >> s;
            return s;
        }
    };

    /**
     * @class dSFMT
     * @brief DSFMT generator class used for dynamic creation
     *
     * This class is one of the main class of DSFMT dynamic creator.
     * This class is designed to be called from programs in MTToolBox,
     * but is not a subclass of some abstract class.
     * Instead, this class is passed to them as template parameters.
     */
    class dSFMT : public ReducibleGenerator<w128_t> {
    public:
        /**
         * Constructor by mexp.
         * @param mexp Mersenne Exponent
         */
        dSFMT(int mexp) {
            size = (mexp - 128) / 104 + 1;
            state = new w128_t[size];
            param.mexp = mexp;
            param.pos1 = 0;
            param.sl1 = 0;
            param.msk1 = 0;
            param.msk2 = 0;
            param.parity1 = 0;
            param.parity2 = 0;
            index = 0;
            start_mode = 0;
            weight_mode = 2;
            previous.u64[0] = 0;
            previous.u64[1] = 0;
            lung.u64[0] = 0;
            lung.u64[1] = 0;
            prefix = 0;
            fixedSL1 = -1;
            fixedPOS1 = -1;
        }

        ~dSFMT() {
            delete[] state;
        }

        /**
         * The copy constructor.
         * @param src The origin of copy.
         */
        dSFMT(const dSFMT& src) : param(src.param) {
            size = src.size;
            state = new w128_t[size];
            for (int i = 0; i < size; i++) {
                state[i] = src.state[i];
            }
            lung = src.lung;
            index = src.index;
            start_mode = src.start_mode;
            weight_mode = src.weight_mode;
            previous = src.previous;
            prefix = src.prefix;
            fixedSL1 = src.fixedSL1;
            fixedPOS1 = src.fixedPOS1;
        }

        /**
         * Constructor by parameter.
         * @param src_param
         */
        dSFMT(const dSFMT_param& src_param) : param(src_param) {
            size = (src_param.mexp - 128) / 104 + 1;
            state = new w128_t[size];
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < 2; j++) {
                    state[i].u64[j] = 0;
                }
            }
            index = 0;
            start_mode = 0;
            weight_mode = 2;
            previous.u64[0] = 0;
            previous.u64[1] = 0;
            lung.u64[0] = 0;
            lung.u64[1] = 0;
            prefix = 0;
            fixedSL1 = -1;
            fixedPOS1 = -1;
        }

        EquidistributionCalculatable<w128_t> * clone() const {
            return new dSFMT(*this);
        }

        /**
         * This method initialize internal state.
         * This initialization is simple.
         * @param seed seed for initialization
         */
        void seed(w128_t seed) {
            //setZero();
            uint64_t * pstate = new uint64_t[(size + 1) * 2];
            for (int i = 0; i < (size + 1) * 2; i++) {
                pstate[i] = 0;
            }
            pstate[0] = seed.u64[0];
            pstate[1] = seed.u64[1];
            for (int i = 1; i < (size + 1) * 2; i++) {
                pstate[i] ^= i + UINT64_C(6364136223846793005)
                    * (pstate[i - 1] ^ (pstate[i - 1] >> 62));
            }
            for (int i = 0; i < size; i++) {
                state[i].u64[0] = pstate[i * 2];
                state[i].u64[1] = pstate[i * 2 + 1];
            }
            lung.u64[0] = pstate[size * 2];
            lung.u64[1] = pstate[size * 2 + 1];
            delete[] pstate;
            index = 0;
            previous.u64[0] = 0;
            previous.u64[1] = 0;
            setup_prefix();
        }

        void do_recursion(w128_t *r, w128_t *a, w128_t *b, w128_t *lung) {
            uint64_t t0;
            uint64_t t1;
            uint64_t L0;
            uint64_t L1;

            t0 = a->u64[0];
            t1 = a->u64[1];
            L0 = lung->u64[0];
            L1 = lung->u64[1];
            lung->u64[0] = (t0 << param.sl1) ^ (L1 >> 32) ^ (L1 << 32)
                ^ b->u64[0];
            lung->u64[1] = (t1 << param.sl1) ^ (L0 >> 32) ^ (L0 << 32)
                ^ b->u64[1];
            r->u64[0] = (lung->u64[0] >> sr1)
                ^ (lung->u64[0] & param.msk1) ^ t0;
            r->u64[1] = (lung->u64[1] >> sr1)
                ^ (lung->u64[1] & param.msk2) ^ t1;
        }

        /**
         * Important state transition function.
         */
        void next_state() {
            index = (index + 1) % size;
            do_recursion(&state[index],
                         &state[index],
                         &state[(index + param.pos1) % size],
                         &lung);
        }

        /**
         * Important method, generate new random number
         * @return new pseudo random number
         */
        w128_t generate() {
            next_state();
            w128_t r;
            index = index % size;
            int p = (index + size - 1) % size;
            switch (start_mode) {
            case 0:
                r.u64[0] = state[index].u64[0];
                r.u64[1] = state[index].u64[1];
                break;
            case 1:
            default:
                r.u64[0] = state[p].u64[1];
                r.u64[1] = state[index].u64[0];
                break;
            }
#if defined(DEBUG) && 0
            if (start_mode != 0) {
                cout << "start_mode = " << dec << start_mode
                     << " index = " << dec << index
                     << " p = " << dec << p << endl;
            }
#endif
            w128_t r2;
            switch (weight_mode) {
            case 2:
                r2 = r;
                break;
            case 1:
            default:
                r2.u64[0] = r.u64[0];
                r2.u64[1] = previous.u64[1];
                break;
            }
#if defined(DEBUG) && 0
            cout << "w:" << dec << weight_mode
                 << " s:" << dec << start_mode << endl;
            cout << "r:" << hex << r << endl;
            cout << "p:" << hex << previous << endl;
            cout << "2:" << hex << r2 << endl;
#endif
            previous = r;
            return r2;
        }

        void setStartMode(int mode) {
            start_mode = mode;
        }

        int getStartMode() {
            return start_mode;
        }

        void setWeightMode(int mode) {
            weight_mode = mode;
            previous.u64[0] = 0;
            previous.u64[1] = 0;
        }

        int getWeightMode() {
            return weight_mode;
        }
        /**
         * This method is called by the functions in simple_shortest_basis.hpp
         * This method returns \b bit_len bits of MSB of generated numbers
         * If reverse_bit_flag is set, bits are taken from LSB
         * @param bit_len bit length from MSB or LSB
         * @return generated numbers of bit_len
         */
        // これは呼ばれない
        w128_t generate(int bit_len) {
            w128_t w;
#if 0
            if (reverse_bit_flag) {
                w = reverse_bit(generate());
            } else {
                w = generate();
            }
#endif
            w = generate();
            w128_t mask = make_msb_mask(bit_len);
            return and_mask(w, mask);
        }

        /**
         * make parameters from given sequential number and
         * internal id
         * @param num sequential number
         */
        void setUpParam(ParameterGenerator& mt) {
            if (fixedPOS1 > 0) {
                param.pos1 = fixedPOS1;
            } else {
                param.pos1 = mt.getUint64() % (size - 2) + 1;
            }
            if (fixedSL1 > 0) {
                param.sl1 = fixedSL1;
            } else {
                param.sl1 = mt.getUint64() % (52 - 1) + 1;
            }
            param.msk1 = mt.getUint64() | mt.getUint64();
            param.msk2 = mt.getUint64() | mt.getUint64();
            param.msk1 &= UINT64_C(0x000fffffffffffff);
            param.msk2 &= UINT64_C(0x000fffffffffffff);
        }

        void setZero() {
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < 2; j++) {
                    state[i].u64[j] = 0;
                }
            }
            lung.u64[0] = 0;
            lung.u64[1] = 0;
            index = 0;
            previous.u64[0] = 0;
            previous.u64[1] = 0;
        }

        /**
         * This method is called by the functions in the file
         * simple_shortest_basis.hpp
         * @return true if all elements of state is zero
         */
        bool isZero() const {
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < 2; j++) {
                    if (state[i].u64[j] != 0) {
                        return false;
                    }
                }
            }
            if ((lung.u64[0] != 0) || (lung.u64[1] != 0)) {
                return false;
            }
            switch (weight_mode) {
            case 2:
                return true;
            case 1:
            default:
                return previous.u[1] == 0;
            }
        }

        void setParityValue(w128_t parity) {
            lung = parity;
            param.parity1 = parity.u64[0];
            param.parity2 = parity.u64[1];
        }

        w128_t getParityValue() const {
            return lung;
        }

        void setOneBit(int bitPos) {
            setZero();
            if (bitPos < size * 104) {
                int idx = bitPos / 104;
                int p = (bitPos / 52) % 2;
                int r = bitPos % 52;
                state[idx].u64[p] = UINT64_C(1) << r;
            } else {
                bitPos = bitPos - size * 104;
                int p = (bitPos / 64) % 2;
                int r = bitPos % 64;
                lung.u64[p] = UINT64_C(1) << r;
            }
        }

        /**
         * This method is called by functions in the file
         * simple_shortest_basis.hpp addition of internal state as
         * GF(2) vector is possible when state transition function and
         * output function is GF(2)-linear.
         * @param that DSFMT generator added to this generator
         */
        void add(EquidistributionCalculatable<w128_t>& other) {
            dSFMT *that = dynamic_cast<dSFMT *>(&other);
            if (that == 0) {
                throw std::invalid_argument(
                    "the adder should have same type as the addee.");
            }
            this->add(that);
        }

        void add(const dSFMT * that) {
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < 2; j++) {
                    state[(i + index) % size].u64[j]
                        ^= that->state[(i + that->index) % size].u64[j];
                }
            }
            lung ^= that->lung;
            previous ^= that->previous;
        }

        int getMexp() const {
            return param.mexp;
        }

        int bitSize() const {
            return size * 104 + 128;
        }

        const std::string getHeaderString() {
            return param.get_header();
        }

        const std::string getParamString() {
            return param.get_string();
        }

        /**
         * This method is called by the functions in search_temper.hpp
         * to calculate the equidistribution properties from LSB
         */
        void set_reverse_bit() {
            //reverse_bit_flag = true;
        }

        /**
         * This method is called by the functions in search_temper.hpp
         * to reset the reverse_bit_flag
         */
        void reset_reverse_bit() {
            //reverse_bit_flag = false;
        }

        void setPrefix(uint64_t p) {
            prefix = p;
        }

        void setConst() {
            const uint64_t high = UINT64_C(0x3ff0000000000000);
            dSFMT tmp(*this);
            tmp.setZero();
            tmp.setPrefix(high);
            tmp.setup_prefix();
            tmp.next_state();
            setZero();
            setPrefix(high);
            setup_prefix();
            add(tmp);
            setPrefix(0);
        }

        bool equals(const dSFMT& that) {
            if ((lung.u64[0] != that.lung.u64[0])
                || (lung.u64[1] != that.lung.u64[1])) {
                return false;
            }
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < 2; j++) {
                    if (state[(index + i) % size].u64[j]
                        != that.state[(that.index + i) % size].u64[j]) {
                        return false;
                    }
                }
            }
            return true;
        }

        void setFixPoint(const w128_t& fix) {
            param.fix1 = fix.u64[0];
            param.fix2 = fix.u64[1];
        }

        void d_p() {
            cout << "index = " << dec << index << endl;
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < 2; j++) {
                    cout << setfill('0') << setw(16) << hex << state[i].u64[j]
                         << " ";
                }
                cout << endl;
            }
            for (int j = 0; j < 2; j++) {
                cout << setfill('0') << setw(16) << hex << lung.u64[j]
                     << " ";
            }
            cout << endl;
        }

        int periodCertification(bool noFix = false) {
            w128_t tmp;
            w128_t parity;
            parity.u64[0] = param.parity1;
            parity.u64[1] = param.parity2;
            w128_t fix;
            fix.u64[0] = param.fix1;
            fix.u64[1] = param.fix2;
            if (noFix) {
                tmp = lung & parity;
            } else {
                tmp = lung ^ fix;
                tmp = tmp & parity;
            }
            int c = count_bit(tmp.u64[0]);
            c += count_bit(tmp.u64[1]);
            c &= 1;
            if (c == 1) {
                return 1;
            }
            if ((parity.u64[1] & 1) == 1) {
                lung.u64[1] ^= 1;
                return 0;
            }
            for (int i = 1; i >= 0; i--) {
                uint64_t work = 1;
                for (int j = 0; j < 64; j++) {
                    if ((work & parity.u64[i]) != 0) {
                        lung.u64[i] ^= work;
                        return 0;
                    }
                    work = work << 1;
                }
            }
            return 0;
        }
        void setFixedSL1(int value) {
            fixedSL1 = value;
        }
        void setFixedPOS1(int value) {
            fixedPOS1 = value;
        }
    private:
        dSFMT& operator=(const dSFMT&) {
            throw std::logic_error("can't assign");
        }
        void setup_prefix() {
            const uint64_t clear = UINT64_C(0x000fffffffffffff);
            for (int i = 0; i < size; i++) {
                state[i].u64[0] &= clear;
                state[i].u64[1] &= clear;
            }
            if (prefix != 0) {
                for (int i = 0; i < size; i++) {
                    state[i].u64[0] |= prefix;
                    state[i].u64[1] |= prefix;
                }
            }
        }
        enum {sr1 = 12};
        int fixedSL1;
        int fixedPOS1;
        int size;
        int index;
        int start_mode;
        int weight_mode;
        w128_t * state;
        w128_t lung;
        dSFMT_param param;
        w128_t previous;
        uint64_t prefix;
    };
}

#endif


src/dSFMTdc.cpp
/**
 * @file dSFMTdc.cpp
 *
 * @brief The main function of parameter generator of dSFMT.
 *
 * The functions in this file are simple. They parse the command line
 * options and call all_in_one function which does almost all things.
 * Users can change this file so that it fits to their applications
 * and OS.
 *
 * @author Mutsuo Saito (Manieth corp.)
 * @author Makoto Matsumoto (Hiroshima University)
 *
 * Copyright (C) 2015 Mutsuo Saito, Makoto Matsumoto,
 * Hiroshima University.
 * All rights reserved.
 *
 * The 3-clause BSD License is applied to this software, see
 * LICENSE.txt
 */
#include <stdint.h>
#include <inttypes.h>
#include <time.h>
#include <fcntl.h>
#include <unistd.h>
#include <string>
#include <sstream>
#include <fstream>
#include <MTToolBox/AlgorithmReducibleRecursionSearch.hpp>
#include <MTToolBox/AlgorithmCalculateParity.hpp>
#include <MTToolBox/AlgorithmEquidistribution.hpp>
#include <MTToolBox/MersenneTwister64.hpp>
#include <NTL/GF2X.h>
#include <getopt.h>
#include "dSFMTsearch.hpp"
#include "AlgorithmDSFMTEquidistribution.hpp"
#include "Annihilate.h"
#include "calc_fixpoint.h"
#include "mpicontrol.hpp"

using namespace std;
using namespace MTToolBox;
using namespace NTL;

class options {
public:
    int mexp;
    bool verbose;
    int fixedSL1;
    int fixedPOS1;
    uint64_t seed;
    std::string filename;
    long count;
};

bool parse_opt(options& opt, int argc, char **argv);


int search(options& opt, ostream& os, int count);

/**
 * parse command line option, and search parameters
 * @param argc number of arguments
 * @param argv value of arguments
 * @return 0 if this ends normally
 */
int main(int argc, char** argv) {
    MPIControl mpi(&argc, &argv);
    options opt;
    bool parse = parse_opt(opt, argc, argv);
    if (!parse) {
        return -1;
    }
    ofstream ofs;
// MPI OUTPUT CHANGE
#if !defined(NO_MPI)
    // MPI では .txt を付けずに指定する。それが一番簡単
    opt.seed = opt.seed + mpi.getRank();
    if (!opt.filename.empty()) {
        char buff[200];
        sprintf(buff, ".s%04ld-%03d.txt", opt.seed, mpi.getRank());
        opt.filename += buff;
    }
#endif
    if (!opt.filename.empty()) {
        ofs.open(opt.filename.c_str());
        if (!ofs) {
            cerr << "can't open file:" << opt.filename << endl;
            return -1;
        }
        return search(opt, ofs, opt.count);
    } else {
        return search(opt, cout, opt.count);
    }
}

/**
 * search parameters using all_in_one function in the file search_all.hpp
 * @param opt command line options
 * @param count number of parameters user requested
 * @return 0 if this ends normally
 */
int search(options& opt, ostream& os, int count) {
    MersenneTwister64 mt(opt.seed);
    dSFMT g(opt.mexp);

    os << "seed = " << dec << opt.seed << endl;
    if (opt.verbose) {
        time_t t = time(NULL);
        os << "search start at " << ctime(&t);
    }
    if (opt.fixedSL1 > 0) {
        g.setFixedSL1(opt.fixedSL1);
    }
    if (opt.fixedPOS1 > 0) {
        g.setFixedPOS1(opt.fixedPOS1);
    }
    AlgorithmReducibleRecursionSearch<w128_t> ars(g, mt);
    int i = 0;
    AlgorithmCalculateParity<w128_t, dSFMT> cp;
    os << "# " << g.getHeaderString() << ", delta52"
         << endl;
    while (i < count) {
        if (ars.start(opt.mexp * 100)) {
            GF2X irreducible = ars.getIrreducibleFactor();
            GF2X characteristic = ars.getCharacteristicPolynomial();
            //os << "deg irreducible = " << dec << deg(irreducible) << endl;
            //os << "deg characteristic = " << dec << deg(characteristic)
            //     << endl;
            //os << "deg quotient = " << dec << deg(quotient) << endl;
            if (deg(irreducible) != opt.mexp) {
                os << "error" << endl;
                return -1;
            }
            getLCMPoly(characteristic, g);
            GF2X quotient = characteristic / irreducible;
            w128_t fixpoint = calc_fixpoint(g, irreducible, quotient);
            g.setFixPoint(fixpoint);
            cp.searchParity(g, irreducible);
            w128_t seed = {{1, 0, 0, 0}};
            g.seed(seed);
            if (!anni(g)) {
                return -1;
            }
            annihilate<w128_t>(&g, quotient);
            int veq52[52];
            DSFMTInfo info;
            info.bitSize = 128;
            info.elementNo = 2;
            int delta52
                = calc_dSFMT_equidistribution<w128_t, dSFMT>(g, veq52, 52, info,
                                                           opt.mexp);
            os << g.getParamString();
            os << dec << delta52 << endl;
            i++;
        } else {
            os << "search failed" << endl;
            break;
        }
    }
    if (opt.verbose) {
        time_t t = time(NULL);
        os << "search end at " << ctime(&t) << endl;
    }
    return 0;
}

static void output_help(string& pgm);

/**
 * command line option parser
 * @param opt a structure to keep the result of parsing
 * @param argc number of command line arguments
 * @param argv command line arguments
 * @param start default start value
 * @return command line options have error, or not
 */
bool parse_opt(options& opt, int argc, char **argv) {
    opt.verbose = false;
    opt.mexp = 0;
    opt.count = 1;
    opt.seed = (uint64_t)clock();
    opt.filename = "";
    opt.fixedSL1 = -1;
    opt.fixedPOS1 = -1;
    int c;
    bool error = false;
    string pgm = argv[0];
    static struct option longopts[] = {
        {"verbose", no_argument, NULL, 'v'},
        {"file", required_argument, NULL, 'f'},
        {"count", required_argument, NULL, 'c'},
        {"seed", required_argument, NULL, 's'},
        {"fixed-sl1", required_argument, NULL, 'x'},
        {"fixed-pos1", required_argument, NULL, 'X'},
        {NULL, 0, NULL, 0}};
    errno = 0;
    for (;;) {
        c = getopt_long(argc, argv, "vs:f:c:x:X:", longopts, NULL);
        if (error) {
            break;
        }
        if (c == -1) {
            break;
        }
        switch (c) {
        case 's':
            opt.seed = strtoull(optarg, NULL, 0);
            if (errno) {
                error = true;
                cerr << "seed must be a number" << endl;
            }
            break;
        case 'x':
            opt.fixedSL1 = strtoull(optarg, NULL, 0);
            if (errno) {
                error = true;
                cerr << "fixed sl1 must be a number" << endl;
            }
            break;
        case 'X':
            opt.fixedPOS1 = strtoull(optarg, NULL, 0);
            if (errno) {
                error = true;
                cerr << "fixed pos1 must be a number" << endl;
            }
            break;
        case 'v':
            opt.verbose = true;
            break;
        case 'f':
            opt.filename = optarg;
            break;
        case 'c':
            opt.count = strtoll(optarg, NULL, 10);
            if (errno) {
                error = true;
                cerr << "count must be a number" << endl;
            }
            break;
        case '?':
        default:
            error = true;
            break;
        }
    }
    argc -= optind;
    argv += optind;
    if (argc < 1) {
        error = true;
    } else {
        long mexp = strtol(argv[0], NULL, 10);
        static const int allowed_mexp[] = {521, 1279,
                                           2203, 4253,
                                           11213, 19937,
                                           44497, -1};
        if (! errno) {
            bool found = false;
            for (int i = 0; allowed_mexp[i] > 0; i++) {
                if (mexp == allowed_mexp[i]) {
                    found = true;
                    break;
                }
            }
            if (! found) {
                error = true;
            }
        }
        if (errno || error){
            error = true;
            cerr << "mexp must be one of ";
            for (int i = 0; allowed_mexp[i] > 0; i++) {
                cerr << dec << allowed_mexp[i] << " ";
            }
            cerr << endl;
        }
        opt.mexp = mexp;
    }
#if 0
    if (!opt.filename.empty()) {
        ofstream ofs(opt.filename.c_str());
        if (ofs) {
            ofs.close();
        } else {
            error = true;
            cerr << "can't open file:" << opt.filename << endl;
        }
    }
#endif
    if (error) {
        output_help(pgm);
        return false;
    }
    return true;
}

/**
 * showing help message
 * @param pgm program name
 */
static void output_help(string& pgm) {
    cerr << "usage:" << endl;
    cerr << pgm
         << " [-s seed] [-v] [-c count]"
         << " [-f outputfile]"
         << " mexp"
         << endl;
    static string help_string1 = "\n"
"--verbose, -v        Verbose mode. Output parameters, calculation time, etc.\n"
"--file, -f filename  Parameters are outputted to this file. without this\n"
"                     option, parameters are outputted to standard output.\n"
"--count, -c count    Output count. The number of parameters to be outputted.\n"
"--seed, -s seed      seed of randomness.\n"
"--fixed-sl1          fix the parameter sl1 to given value.\n"
"--fixed-pos1         fix the parameter pos1 to given value.\n"
"mexp                 mersenne exponent.\n"
        ;
    cerr << help_string1 << endl;
}
